c++/src/kj/common-test.c++

```diff
@@ -75,6 +75,37 @@ TEST(Common, Maybe) {
       ADD_FAILURE();
     }
     EXPECT_EQ(123, m.orDefault(456));
+    bool ranLazy = false;
+    EXPECT_EQ(123, m.orDefault([&] {
+      ranLazy = true;
+      return 456;
+    }));
+    EXPECT_FALSE(ranLazy);
+
+    KJ_IF_MAYBE(v, m) {
+      const int& ref = m.orDefault([notUsed = 5]() -> const int& { return notUsed; });
```

Comment:
Hmm... this fails to build on MSVC because for some reason it selects the `&` override instead of `const &` & fails with a hard error instead of continuing with other overloads (but how did the `Result` template parameter even get evaluated?! Should have been a SFINAE failure. It's also a mystery why it says the evaluation of `lazyDefaultValue()` is returning a `const int` when it explicitly returns a `const int&` (i.e. I'd expect the error message to say it can't convert a `const int &` to `int&`

Reply:
I think that choosing the `&` overload is correct. In this case, `decltype(instance<bool>() ? instance<T&>() : instance<F>()())` should find that the branches of the `?:` are `int&` and `const int&`, which it should reconcile by choosing `const int&` as the result type, since that's the common type both can be converted to. That is to say, the following is valid according to GCC and Clang:

```
const int& foo(bool b, int& r1, const int& r2) {
  return b ? r1 : r2;
}
```

But it appears that MSVC is incorrectly choosing `int&` as the result type. That seems like an MSVC bug.

I think we can adjust the test to avoid this case, by either returning a non-const reference here, or returning an rvalue. I don't think people will commonly return references from these callbacks anyway.